src/backends/interflop-vprec/interflop_vprec.c

```diff
@@ -273,34 +378,87 @@ void _set_vprec_inst_mode(vprec_inst_mode mode) {
 // Round the float with the given precision
 static float _vprec_round_binary32(float a, char is_input, void *context,
                                    int binary32_range, int binary32_precision) {
+  t_context *currentContext = (t_context *)context;
+
+  /* test if 'a' is a special case */
   if (!isfinite(a)) {
     return a;
   }
 
-  /* round to zero or set to infinity if underflow or overflow compare to
+  /* round to zero or set to infinity if underflow or overflow compared to
    * VPRECLIB_BINARY32_RANGE */
   int emax = (1 << (binary32_range - 1)) - 1;
-  // here emin is the smallest exponent in the *normal* range
+  /* here emin is the smallest exponent in the *normal* range */
   int emin = 1 - emax;
 
+  // /* in absolute error mode, the error threshold also gives the possible
+  //  * underflow limit */
+  // if (currentContext->absErr == true) {
+  //   if (currentContext->relErr == true) {
+  //     /* relative and absolute error mode */
+  //     if (currentContext->absErr_exp > emin)
+  //       emin = currentContext->absErr_exp;
+  //   } else {
+  //     /* absolute error mode */
+  //     emin = currentContext->absErr_exp;
+  //   }
+  // }
+
   binary32 aexp = {.f32 = a};
   aexp.s32 = ((FLOAT_GET_EXP & aexp.u32) >> FLOAT_PMAN_SIZE) - FLOAT_EXP_COMP;
 
-  /* check for overflow or underflow in target range */
+  /* check for overflow in target range */
   if (aexp.s32 > emax) {
     a = a * INFINITY;
     return a;
   }
 
+  /* check for underflow in target range */
   if (aexp.s32 < emin) {
-    if ((((t_context *)context)->daz && is_input) ||
-        (((t_context *)context)->ftz && !is_input)) {
+    /* underflow case: possibly a denormal */
+    if ((currentContext->daz && is_input) ||
+        (currentContext->ftz && !is_input)) {
       a = 0;
     } else {
-      a = handle_binary32_denormal(a, emin, binary32_precision);
+      if (currentContext->absErr == true) {
+        /* absolute error, or absolute and relative error mode */
+        int binary32_precision_adjusted;
+        binary32_precision_adjusted =
+            compute_absErr_vprec_binary32(true, context, 0, binary32_precision);
+        a = handle_binary32_denormal(a, emin, binary32_precision_adjusted);
+      } else {
+        /* relative error mode */
+        a = handle_binary32_denormal(a, emin, binary32_precision);
+      }
     }
   } else {
-    a = round_binary32_normal(a, binary32_precision);
+    /* else, normal case: can be executed even if a
+     previously rounded and truncated as denormal */
+    if (currentContext->absErr == true) {
```

Comment:
This block and the counterpart for 64 below should be factored in one or two separate functions.

Reply:
This block has been factorized into the handle_binary32_normal_absErr and handle_binary64_normal_absErr functions.